### PR TITLE
Feat/notification logic

### LINKED ISSUE #7721: Notifications logic
We already have notification redux logic in common. Now we need to somehow show notifications added to redux store. 

1. We will need some `NotificationRenderer` component that will be subscribed to redux state and show notifications in UI. This rerender will be placed somewhere in app root.

### PR DESCRIPTION
**Pull request by PeKne**
Transaction notifications are displayed in reaction on @suite-common/toast-notification redux state.

Tested on on-chain Bitcoin-Testnet transactions.

## Related Issue

Resolve #7721

## Screenshots:

https://user-images.githubusercontent.com/26143964/223060601-0343d349-5f55-4645-bef5-a90de3728692.mov

### CODE REVIEW

**Comment by PeKne** [comment]
Confirmed transaction notification added:
Screenshot 2023-03-08 at 15 09 05

File: suite-common/wallet-core/src/accounts/accountsReducer.ts [hunk 1]
```diff
@@ -194,6 +194,18 @@ export const selectAccountByDescriptorAndNetworkSymbol = memoizeWithArgs(
...
 
+export const selectAccountKeyByDescriptorAndNetworkSymbol = memoizeWithArgs(
+    (state: AccountsRootState, accountDescriptor: string, networkSymbol: NetworkSymbol) => {
+        const account = selectAccountByDescriptorAndNetworkSymbol(
+            state,
+            accountDescriptor,
+            networkSymbol,
+        );
+
+        return account?.key;
```
Comment (juriczech) [inline suite-common/wallet-core/src/accounts/accountsReducer.ts]: nit: I'd return `null` instead of undefined

Comment (juriczech) [reply suite-common/wallet-core/src/accounts/accountsReducer.ts]: Also I think it's quite nice to type return value from selectors. Even though ts can see what the return value is based on code changes, it is IMO good to "lock" that value with return types.

File: suite-common/wallet-core/src/accounts/accountsReducer.ts [hunk 2]
```diff
@@ -194,6 +194,18 @@ export const selectAccountByDescriptorAndNetworkSymbol = memoizeWithArgs(
     },
 );
 
+export const selectAccountKeyByDescriptorAndNetworkSymbol = memoizeWithArgs(
+    (state: AccountsRootState, accountDescriptor: string, networkSymbol: NetworkSymbol) => {
```
Comment (juriczech) [inline suite-common/wallet-core/src/accounts/accountsReducer.ts]:
```suggestion
    (state: AccountsRootState, accountDescriptor: string, networkSymbol: NetworkSymbol): AccountKey | null => {
```

File: suite-native/module-accounts/src/components/TransactionListHeader.tsx [hunk 3]
```diff
@@ -23,9 +25,14 @@ type AccountDetailHeaderProps = {
     accountKey: AccountKey;
 };
 
+type AccountDetailNavigationProps = TabToStackCompositeNavigationProp<
```
Comment (juriczech) [inline suite-native/module-accounts/src/components/TransactionListHeader.tsx]: nit: why account detail?

File: suite-native/module-dev-utils/src/screens/DemoScreen.tsx [hunk 4]
```diff
@@ -80,7 +80,8 @@ export const DemoScreen = () => {
                     >
                         {transactionNotifications.map(notification => (
                             <TransactionNotification
-                                {...notification}
+                                // @ts-expect-error
```
Comment (juriczech) [inline suite-native/module-dev-utils/src/screens/DemoScreen.tsx]: Shouldn't be needed. Just type `transactionNotifications` and get rid of this expect-error.

File: suite-native/module-send-receive/src/screens/ReceiveAccountsScreen.tsx [hunk 5]
```diff
@@ -1,33 +1,22 @@
...
-    );
-};
+export const ReceiveAccountsScreen = ({
+    navigation,
+}: StackProps<SendReceiveStackParamList, SendReceiveStackRoutes.ReceiveAccounts>) => (
+    <Screen header={<ScreenHeader title="Receive to" hasGoBackIcon={false} />}>
+        <AccountsList
+            onSelectAccount={accountKey =>
+                navigation.navigate(SendReceiveStackRoutes.Receive, { accountKey })
+            }
```
Comment (juriczech) [inline suite-native/module-send-receive/src/screens/ReceiveAccountsScreen.tsx]: nit: Can you please move this to handler function? I know it's been like that but small improvement

File: suite-native/notifications/src/components/TransactionNotification.tsx [hunk 6]
```diff
@@ -1,92 +1,122 @@
...
         useNavigation<StackNavigationProps<RootStackParamList, RootStackRoutes.AppTabs>>();
+    const dispatch = useDispatch();
+    const accountKey =
+        useSelector((state: AccountsRootState) =>
+            selectAccountKeyByDescriptorAndNetworkSymbol(
+                state,
+                notification.descriptor,
+                notification.symbol,
+            ),
+        ) ?? '';
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotification.tsx]: You shouldn't do fallbacks like this in component. This should be handled in selector itself

File: suite-native/notifications/src/components/TransactionNotification.tsx [hunk 7]
```diff
@@ -1,92 +1,122 @@
...
+    TransactionEventNotification,
+} from '@suite-common/toast-notifications';
 
 import { Notification } from './Notification';
 import { TransactionNotificationDescription } from './TransactionNotificationDescription';
 
 type TransactionNotificationProps = {
-    txid: string;
-    accountKey: AccountKey;
+    notification: TransactionEventNotification;
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotification.tsx]: nit: better name would be transactionNotification because it's weird when down below you access notification txid. I would like to have the transaction in naming.

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx [hunk 8]
```diff
@@ -17,24 +15,16 @@ const addressContainerStyle = prepareNativeStyle(_ => ({
...
-            <CryptoAmountFormatter
-                value={amount}
-                network={networkSymbol}
-                isBalance={false}
-                color="textSubdued"
-                variant="label"
-            />
             <Text color="textSubdued" variant="label">
-                {` ${prefix} `}
+                {`${formattedAmount} ${prefix} `}
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotificationDescription.tsx]:
```suggestion
                {formattedAmount} {prefix}
```

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx [hunk 9]
```diff
@@ -17,24 +15,16 @@ const addressContainerStyle = prepareNativeStyle(_ => ({
...
 
     return (
         <Box flexDirection="row">
-            <CryptoAmountFormatter
-                value={amount}
-                network={networkSymbol}
-                isBalance={false}
-                color="textSubdued"
-                variant="label"
-            />
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotificationDescription.tsx]: Why this change? If you want to dispatch notification for new transaction, you won't have formatted amount there and should be handled here.

File: suite-native/notifications/src/components/TransactionNotification.tsx [hunk 10]
```diff
@@ -1,92 +1,122 @@
...
 
-    const transaction = useSelector((state: TransactionsRootState) =>
-        selectTransactionByTxidAndAccountKey(state, txid, accountKey),
+    const transactionTargetAddress = useSelector((state: TransactionsRootState) =>
+        selectTransactionFirstTargetAddress(state, notification.txid, accountKey),
     );
 
-    const transactionTargetAddress = useSelector((state: TransactionsRootState) =>
-        selectTransactionFirstTargetAddress(state, txid, accountKey),
+    const removeNotification = useCallback(
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotification.tsx]: nit: no need for use callback since notification is not using memo

**Review by juriczech — CHANGES_REQUESTED** [review]

File: suite-common/wallet-core/src/accounts/accountsReducer.ts [hunk 11]
```diff
@@ -194,6 +194,18 @@ export const selectAccountByDescriptorAndNetworkSymbol = memoizeWithArgs(
     },
 );
 
+export const selectAccountKeyByDescriptorAndNetworkSymbol = memoizeWithArgs(
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/accounts/accountsReducer.ts:197]: do we need this? Also why we can't use key directly?

File: suite-common/toast-notifications/src/types.ts [hunk 12]
```diff
@@ -166,3 +166,14 @@ export type NotificationsState = NotificationEntry[];
 export type NotificationsRootState = {
     notifications: NotificationsState;
 };
+
+export type TransactionEventNotification = {
```
Comment (Nodonisko) [inline suite-common/toast-notifications/src/types.ts]: Why we need new type? I guess desktop suite already had notifications about transactions so everything should be already here.

File: suite-common/wallet-core/src/accounts/accountsReducer.ts (hunk 1 above)
Comment (Nodonisko) [reply suite-common/wallet-core/src/accounts/accountsReducer.ts]: I doesn't matter basically both `undefined` or `null` sucks.  These cases are exactly places where you should use `Result` from ts-belt of fp-ts

File: suite-native/module-dev-utils/src/fixtures.ts [hunk 13]
```diff
@@ -1,12 +1,24 @@
 export const transactionNotifications = [
     {
-        txid: 'dfdc5af5fae8fd0681907a32bab361e356535be70bacebb64586b23a38bd5a5e',
```
Comment (Nodonisko) [inline suite-native/module-dev-utils/src/fixtures.ts:3]: Why so many extra field added? Previously only `txid` and `accountKey` was here and that should enough to select everything you need using selectors.

File: suite-native/navigation/src/navigators.ts [hunk 14]
```diff
@@ -37,10 +38,15 @@ export type AccountsStackParamList = {
     [AccountsStackRoutes.AccountDetailSettings]: { accountKey: string };
 };
 
+export type SendReceiveStackParamList = {
+    [SendReceiveStackRoutes.ReceiveAccounts]: undefined;
+    [SendReceiveStackRoutes.Receive]: { accountKey: string };
```
Comment (Nodonisko) [inline suite-native/navigation/src/navigators.ts]: We should have AccountKey type somewhere.

File: suite-native/notifications/src/components/Notification.tsx [hunk 15]
```diff
@@ -86,23 +87,38 @@ export const Notification = ({
         ],
     }));
 
-    if (isHidden) return null;
+    const enteringAnimation = (_: EntryAnimationsValues) => {
```
Comment (Nodonisko) [inline suite-native/notifications/src/components/Notification.tsx]: `handleEnteringAnimation`

File: suite-native/notifications/src/components/NotificationRenderer.tsx [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        flexDirection: 'row',
+        marginTop: topSafeAreaInset,
+        paddingHorizontal: utils.spacings.medium,
+    }),
+);
+
+export const NotificationRenderer = ({ children }: NotificationRendererProps) => {
+    const { applyStyle } = useNativeStyles();
+    const { top: topSafeAreaInset } = useSafeAreaInsets();
+    const transactionNotifications = useSelector((state: NotificationsRootState) =>
```
Comment (Nodonisko) [inline suite-native/notifications/src/components/NotificationRenderer.tsx]: if you don't have extra params you should be able to do just `useSelector(selectTransactionNotifications)`

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 10 above)
Comment (Nodonisko) [inline suite-native/notifications/src/components/TransactionNotification.tsx]: `handleRemoveNotification`

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx [hunk 17]
```diff
@@ -17,24 +15,16 @@ const addressContainerStyle = prepareNativeStyle(_ => ({
...
+    formattedAmount,
     prefix,
     targetAddress,
-    networkSymbol,
 }: TransactionNotificationDescriptionProps) => {
     const { applyStyle } = useNativeStyles();
 
     return (
         <Box flexDirection="row">
-            <CryptoAmountFormatter
```
Comment (Nodonisko) [inline suite-native/notifications/src/components/TransactionNotificationDescription.tsx]: Never even use `formattedAmount`. That is nonsense from desktop Suite which is bad for many reasons. Always pass raw value and only format it using formatter. Here you should use `amount` field from `Transaction`.

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/wallet-core/src/accounts/accountsReducer.ts (hunk 1 above)
Comment (PeKne) [reply suite-common/wallet-core/src/accounts/accountsReducer.ts]: So should I change it then?

Comment (juriczech) [reply suite-common/wallet-core/src/accounts/accountsReducer.ts]: @Nodonisko Yeah, it won't change anything, just semantic thing. Feel free to resolve if you disagree

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx (hunk 8 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotificationDescription.tsx]: Same as the last time, the trailing whitespace is needed for correct formatting.

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 10 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: The `notification` is a prop why would you memoize a prop?

File: suite-common/wallet-core/src/accounts/accountsReducer.ts (hunk 11 above)
Comment (PeKne) [reply suite-common/wallet-core/src/accounts/accountsReducer.ts:197]: We cannot use the key directly because this information is not inside of `EventNotification` object. The shape is as follows: 
```
{
        context: 'event',
        descriptor:
            'vpub5ZjRPuuMiEQnbwEDi9jtH1FaJMajZW78uZ1t3RJXKhxyMoTnPraKwGxiDo9SguDYvSieqjoLJxW5n2t9156RR1oeqRnURuftNZTzejBc4pa',
        device: undefined,
        formattedAmount: '0.00055 TEST',
        id: 1678029249638,
        symbol: 'test',
        txid: '2a291100058438b66b479f2d7f474a9bc7dd36b7988220af5422a99932938f00',
        type: 'tx-received',
}        
```

for that reason we first need to get accountKey from descriptor and networkSymbol. When we have the accountKey then we are able to get transaction.

The operations then  are:
[descriptor, networkSymbol] -> accountKey
[accountKey, txid] => transaction object

File: suite-native/module-accounts/src/components/TransactionListHeader.tsx (hunk 3 above)
Comment (PeKne) [reply suite-native/module-accounts/src/components/TransactionListHeader.tsx]: because the navigator is in AccountDetail screen. But it should be more general. I changed it to `AccountsNavigationProps` in 3ec0945bf

File: suite-native/module-send-receive/src/screens/ReceiveAccountsScreen.tsx (hunk 5 above)
Comment (PeKne) [reply suite-native/module-send-receive/src/screens/ReceiveAccountsScreen.tsx]: fixed in 3ec0945bf

File: suite-common/toast-notifications/src/types.ts (hunk 12 above)
Comment (PeKne) [reply suite-common/toast-notifications/src/types.ts]: fixed in 53e695678

File: suite-common/wallet-core/src/accounts/accountsReducer.ts (hunk 1 above)
Comment (PeKne) [reply suite-common/wallet-core/src/accounts/accountsReducer.ts]: return type added in 53e695678

File: suite-native/module-dev-utils/src/fixtures.ts (hunk 13 above)
Comment (PeKne) [reply suite-native/module-dev-utils/src/fixtures.ts:3]: refactored in 53e695678

File: suite-native/module-dev-utils/src/screens/DemoScreen.tsx (hunk 4 above)
Comment (PeKne) [reply suite-native/module-dev-utils/src/screens/DemoScreen.tsx]: removed in 53e695678

File: suite-native/navigation/src/navigators.ts (hunk 14 above)
Comment (PeKne) [reply suite-native/navigation/src/navigators.ts]: fixed 53e695678

File: suite-native/notifications/src/components/Notification.tsx (hunk 15 above)
Comment (PeKne) [reply suite-native/notifications/src/components/Notification.tsx]: renamed to `runEnteringAnimation` in 53e695678

File: suite-native/notifications/src/components/NotificationRenderer.tsx (hunk 16 above)
Comment (PeKne) [reply suite-native/notifications/src/components/NotificationRenderer.tsx]: fixed 53e695678

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 7 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: is not relevant after refactor i 53e695678

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 6 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: fixed  53e695678

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx (hunk 17 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotificationDescription.tsx]: Ok, I wasn't aware of this. Refactored in 53e695678

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx (hunk 9 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotificationDescription.tsx]: reverted in 53e695678

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx (hunk 8 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotificationDescription.tsx]: 53e695678

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 10 above)
Comment (juriczech) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: Should be removed as discussed

Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: fixed https://github.com/trezor/trezor-suite/commit/53e695678eff69454abfdd5a904a57601e4312e3

**Review by Nodonisko — APPROVED** [review]

File: suite-native/notifications/src/components/TransactionNotification.tsx [hunk 18]
```diff
@@ -1,92 +1,136 @@
...
-        selectTransactionByTxidAndAccountKey(state, txid, accountKey),
+        selectTransactionByTxidAndAccountKey(state, txid, accountKey ?? ''),
     );
 
     const transactionTargetAddress = useSelector((state: TransactionsRootState) =>
-        selectTransactionFirstTargetAddress(state, txid, accountKey),
+        selectTransactionFirstTargetAddress(state, txid, accountKey ?? ''),
+    );
+
+    const handleRemoveNotification = useCallback(
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotification.tsx]: remove `useCallback`

File: suite-native/notifications/src/components/TransactionNotification.tsx [hunk 19]
```diff
@@ -1,92 +1,136 @@
...
         <Notification
             isHiddenAutomatically={isHiddenAutomatically}
+            onHide={handleRemoveNotification}
             onPress={navigateToTransactionDetail}
             title={title}
             description={
                 <TransactionNotificationDescription
-                    networkSymbol={transaction.symbol}
-                    amount={transaction.amount}
+                    amount={transaction?.amount}
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotification.tsx]: Then you don't have to worry about fallback in description component. Undefined shouldn't be acceptable as prop 

```suggestion
                    amount={transaction?.amount ?? null}
```

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx [hunk 20]
```diff
@@ -6,7 +6,7 @@ import { NetworkSymbol } from '@suite-common/wallet-config';
 import { useNativeStyles, prepareNativeStyle } from '@trezor/styles';
 
 type TransactionNotificationDescriptionProps = {
-    amount: string;
+    amount?: string;
```
Comment (juriczech) [inline suite-native/notifications/src/components/TransactionNotificationDescription.tsx]:
```suggestion
    amount: string | null;
```

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 10 above)
Comment (juriczech) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: Still there

Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: 114b78127

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 18 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: 114b78127

File: suite-native/notifications/src/components/TransactionNotification.tsx (hunk 19 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotification.tsx]: 114b78127

File: suite-native/notifications/src/components/TransactionNotificationDescription.tsx (hunk 20 above)
Comment (PeKne) [reply suite-native/notifications/src/components/TransactionNotificationDescription.tsx]: 114b78127

**Review by juriczech — APPROVED** [review]
